ddl/placement/rule.go

```diff
@@ -50,102 +50,16 @@ type Rule struct {
-	err1, err2 = yaml.UnmarshalStrict([]byte(constr1), &constraintsList1), yaml.UnmarshalStrict([]byte(constr2), &constraintsList2)
-	if err2 != nil {
-		// Common constraints can only be a list.
-		return nil, fmt.Errorf("%w: should be [constraint1, ...] (error %s)", ErrInvalidConstraintsFormat, err2)
-	}
-	if err1 != nil {
-		combineType = 0x01
-		err3 = yaml.UnmarshalStrict([]byte(constr1), &constraintsDict1)
-		if err3 != nil {
-			return nil, fmt.Errorf("%w: should be [constraint1, ...] (error %s), {constraint1: cnt1, ...} (error %s), or any yaml compatible representation", ErrInvalidConstraintsFormat, err1, err3)
-		}
-	}
-	switch combineType {
-	case listAndList:
-		/*
-		 * eg: followerConstraint = ["+zone=sh", "+zone=bj"], constraint = ["+disk=ssd"]
-		 * res: followerConstraint = ["+zone=sh", "+zone=bj", "+disk=ssd"]
-		 */
-		if replicas == 0 {
-			if len(constr1) == 0 {
-				return rules, nil
-			}
-			return rules, fmt.Errorf("%w: should be positive", ErrInvalidConstraintsRelicas)
-		}
-		constraintsList1 = append(constraintsList1, constraintsList2...)
-		labelConstraints, err := NewConstraints(constraintsList1)
-		if err != nil {
-			return rules, err
-		}
-		rules = append(rules, &Rule{
-			Count:       int(replicas),
-			Constraints: labelConstraints,
-		})
-		return rules, nil
-	case dictAndList:
-		/*
-		 * eg: followerConstraint = { '+zone=sh, -zone=bj':2, '+zone=sh':1 }, constraint = ['+disk=ssd']
-		 * res: followerConstraint = { '+zone=sh, -zone=bj, +disk=ssd':2, '+zone=sh, +disk=ssd':1 }
-		 */
-		ruleCnt := 0
-		for labels, cnt := range constraintsDict1 {
-			if cnt <= 0 {
-				return rules, fmt.Errorf("%w: count of labels '%s' should be positive, but got %d", ErrInvalidConstraintsMapcnt, labels, cnt)
-			}
-			ruleCnt += cnt
-		}
-		if replicas == 0 {
-			replicas = uint64(ruleCnt)
-		}
-		if int(replicas) < ruleCnt {
-			return rules, fmt.Errorf("%w: should be larger or equal to the number of total replicas, but REPLICAS=%d < total=%d", ErrInvalidConstraintsRelicas, replicas, ruleCnt)
-		}
-		for labels, cnt := range constraintsDict1 {
-			mergeLabels := append(strings.Split(labels, ","), constraintsList2...)
-			labelConstraints, err := NewConstraints(mergeLabels)
-			if err != nil {
-				return rules, err
-			}
-			rules = append(rules, &Rule{
-				Count:       cnt,
-				Constraints: labelConstraints,
-			})
-		}
-		remain := int(replicas) - ruleCnt
-		if remain > 0 {
-			rules = append(rules, &Rule{
-				Count: remain,
-			})
-		}
-		return rules, nil
-	}
-	// empty
-	return rules, nil
-}
-
 // NewRule constructs *Rule from role, count, and constraints. It is here to
 // consistent the behavior of creating new rules.
 func NewRule(role PeerRoleType, replicas uint64, cnst Constraints) *Rule {
-	return &Rule{Role: role, Count: int(replicas), Constraints: cnst}
+	return &Rule{
+		Role:           role,
+		Count:          int(replicas),
+		Constraints:    cnst,
+		LocationLabels: []string{"region", "zone", "rack", "host"},
```

Comment:
just one question: what this two fields is for?

Reply:
For isolation scores, it is used to compare which schedule is better. Say `region=sh,bj,us`, `2,1,0` will have a lower score `1,1,1` because of the existence of `LocationLaels: []string{..., region, ...}`. Similar arguments apply to `zone,rack,host`.

As for `IsolationLevel`, it means two locations are far enough and will affect schedule behaviors. Refer https://github.com/tikv/pd/blob/740fac18e89fa1754a40f5ec885b5e5b6caa2471/server/schedule/filter/filters.go#L696-L701.

Reply:
thanks